framework/include/cppmicroservices/AnyMap.h

```diff
@@ -0,0 +1,322 @@
+
+    typedef AnyMap::reference reference;
+    typedef AnyMap::pointer pointer;
+
+    typedef Iterator iterator;
+
+    Iterator();
+    Iterator(const iterator& it);
+    ~Iterator();
+
+    Iterator(OIter&& it);
+    Iterator(UOIter&& it, IterType type);
+
+    reference operator* () const;
+    pointer operator-> () const;
+
+    iterator& operator++();
+    iterator operator++(int);
+
+    bool operator==(const iterator& x) const;
+    bool operator!=(const iterator& x) const;
+
+  private:
+
+    OIter const& o_it() const;
+    OIter& o_it();
+    UOIter const& uo_it() const;
+    UOIter& uo_it();
+    UOCIIter const& uoci_it() const;
+    UOCIIter& uoci_it();
+
+    union {
+      char o[sizeof(OIter)];
+      char uo[sizeof(UOIter)];
+      char uoci[sizeof(UOCIIter)];
+    } iter;
+
+  };
+
+  typedef Iterator iterator;
+  typedef ConstIterator const_iterator;
+
+  AnyMap(Type type);
+  AnyMap(const OrderedMap& m);
+  AnyMap(const UnorderedMap& m);
+  AnyMap(const UnorderedMapCaseInsensitiveKeys& m);
+
+  AnyMap(const AnyMap& m);
+  AnyMap& operator=(const AnyMap& m);
+
+  ~AnyMap();
+
+  Type GetType() const;
+
+  iterator begin();
+  const_iterator begin() const;
+  const_iterator cbegin() const;
+  iterator end();
+  const_iterator end() const;
+  const_iterator cend() const;
+
+  bool empty() const;
+  size_type size() const;
+  size_type count(const key_type& key) const;
+  void clear();
+
+  mapped_type& at(const key_type& key);
+  const mapped_type& at(const key_type& key) const;
+
+  mapped_type& operator[](const key_type& key);
+  mapped_type& operator[](key_type&& key);
+
+  std::pair<iterator, bool> insert(const value_type& value);
+  const_iterator find(const key_type& key) const;
+
+  /**
+   * Get a key's value, using a compound key notation.
+   *
+   * A compound key consists of one or more key names, concatenated with
+   * the '.' (dot) character. Each key except the last requires the referenced
+   * Any object to be of type \c AnyMap or \c std::vector<Any>. Containers
+   * of type \c std::vector<Any> are indexed using 0-based numerical key names.
+   *
+   * For example, a \c AnyMap object holding data of the following layout
+   * \code{.json}
+   * {
+   *   one: 1,
+   *   two: "two",
+   *   three: {
+   *     a: "anton",
+   *     b: [ 3, 8 ]
+   *   }
+   * }
+   * \endcode
+   * can be queried using the following notation:
+   * \code
+   * map.GetDotted("one"); // returns Any(1)
+   * map.GetDotted("three.a"); // returns Any(std::string("anton"))
+   * map.GetDotted("three.b.1"); // returns Any(8)
+   * \endcode
+   *
+   * @param key The key hierachy to query.
+   * @return A reference to the key's value.
+   *
+   * @throws std::invalid_argument if the \c Any value for a given key is not of type \c AnyMap or \c std::vector<Any>.
+   * std::out_of_range if the key is not found or a numerical index would fall out of the range of an \c int type.
+   */
+  mapped_type& GetDotted(const key_type& key);
```

Comment:
Since AnyMap is a Framework defined class, I think using PascalCase uniformly will help with code readability.